Python debug scripts

mbolivar opened the pull request:
TL;DR
=====

This series finishes rewriting the flash and debug scripts in Python, which should help the CMake switch a bit. Since there were a few issues last time, there are more details below.  **JLink on frdm_k64f has an issue here and on master -- just me?**

It also fixes two minor issues with the Python script (in "scripts: zephyr_flash_debug: add missing nrfjprog line" and "scripts: zephyr_flash_debug: fix ARC OPENOCD_EXTRA_INIT").

Casting a wide net of reviewers among build, arch, and SoC people just in case.

Details
=======

Results for testing Python equivalents to shell scripts follow. Linux-specific shell idioms were converted to portable alternatives for macOS and Windows targets, but this series has only been tested end-to-end on Linux.

Flashing
--------------

Not tested on HW (I don't have boards):

- bossa-flash.sh
- esp32.sh
- nios2.sh

Tested that flashing hello world gives console output:

- dfuutil.sh: 96b_carbon,  arduino_101
- pyocd.sh: frdm_k64f, 96b_nitrogen, nrf52_blenano2
- nrf_flash.sh: nrf52840_pca10056
- openocd.sh: disco_l475_iot1

Tested to run without error only (I don't know how to check the console to make sure the right chip was flashed):

- dfuutil.sh: arduino_101_sss, arduino_101_ble

Debugging
---------------

Not tested on HW (I don't have boards, or it's QEMU):

- arc_debugger.sh
- nios2.sh
- xt-gdb.sh
- qemu.sh

Tested that 'debug' connects and allows resetting and running code:

- openocd.sh: disco_l475_iot1
- pyocd.sh: frdm_k64f, 96b_nitrogen, nrf52_blenano2

**Issue**

- jlink.sh: on frdm_k64f, I tried `make -j8 debug BOARD=frdm_k64f OPENSDA_FW=jlink`. This failed, but this doesn't seem to be a regression; I get the same error on master (https://hastebin.com/ijoxuhaxom.sql). Is it just me?

Other tests
-----------

I hacked up the Python and shell scripts on another branch to log their output to files instead (with some filtering to make the shell versus Python diffs less noisy):

https://github.com/mbolivar/zephyr/tree/python-debug-scripts-testing

On that branch, I ran both flash and debug for every Zephyr board samples/hello_world built on (esp32 and xt-sim had issues using the 0.9.2-rc5 SDK) and verified the logs were equivalent.

Future
------

- make the Python script smarter (the first goal is adding DT-awareness for correct flash offsets when relevant)
- make this less messy (use arguments instead of environment variables), maybe clean up the Makefile.board files accordingly (might just punt this in favor of CMake)

Commit:
scripts: zephyr_flash_debug: generalize to support debugging

Replace the 'flash' method with a 'run' method, which takes a command
to run (flash, debug, or debugserver).

Rename the classes involved appropriately, and generalize the factory
interfaces as needed.

Add documentation and theory of ops.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: debug like arc_debugger.sh

Add debug and debugserver support.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: teach runners a client/server abstraction

Several debugging scripts run setsid before executing a server
process, then run GDB with SIGINT ignored.

Relying on setsid is not portable. Add a popen_ignore_int() helper
that provides a portable alternative, and provide a generic
run_server_and_client() in ZephyrBinaryRunner which uses it to
abstract the pattern.

Subsequent patches will use this to implement the 'debug' command.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: make ARC runner portable

Use run_client_and_server() to work across supported platforms.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: refactor openocd runner

Preparatory work for supporting debugging and flashing. No significant
behavioral changes.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: fix ARC OPENOCD_EXTRA_INIT

Though it isn't used by any in-tree Makefile.boards, looking at the
RIOT OS build system, this is meant to be split along lexical
boundaries defined by the shell, not just whitespace.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
build: don't try debugserver if there's no DEBUG_SCRIPT

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: add debug support to openocd

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
build: interpose zephyr_flash_debug.py in debug/debugserver targets

Just as was previously done for flashing, invoke zephyr_flash_debug.py
from the build system debug and debugserver targets by default. The
Python script will fall back on the corresponding shell script if it
doesn't have its own implementation.

The shell script can be used instead, just as with flashing, by
setting USE_ZEPHYR_FLASH_DEBUG_SHELL to any nonempty value.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: refactor pyocd runner

Prep work for adding debug and debugserver support. No significant
functional differences.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: add debug support to pyocd

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: fix Nios II comment

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: improve debug logging

Ensure that the printed commands can be copy/pasted into a shell with
the same semantics.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: refactor Nios II runner

Prep work for adding debug and debugserver support. No significant
functional differences.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: add missing nrfjprog line

The --pinreset line from the shell script is missing.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

galak commented on scripts/support/zephyr_flash_debug.py:
```diff
@@ -177,71 +312,77 @@ def create_from_env(debug):
+        return command == 'flash' and shell_script == 'bossa-flash.sh'
 
-    def create_from_env(debug):
+    def create_from_env(command, debug):
```
Since you are cleaning all things up, can you look at adding in when I did with PR #4502

mbolivar replied:
Sounds good! That class was a pretty primitive one.

galak commented on Makefile:
```diff
@@ -1390,9 +1390,13 @@ flash: FORCE
 endif
 
 ifneq ($(DEBUG_SCRIPT),)
+ifeq ($(USE_ZEPHYR_FLASH_DEBUG_SHELL),)
```
is there a reason not to just use your script always?

mbolivar replied:
With this series, the script is being used by default now for everything (flash, debug, debugserver).

I'm leaving this in as an escape hatch in case there are problems after the merge, the same way as was done when I re-did the flashing scripts in https://github.com/zephyrproject-rtos/zephyr/pull/4304.

My plan is to remove the shell scripts once these get merged and a few days of testing, but I an remove them at the end of this series as well if that's better.

andrewboie replied:
it's straightforward enough to revert a PR if it causes problems.
problem with checks like this, in my experience they don't always get removed and become cruft.

mbolivar replied:
I went ahead and deleted all the shell scripts in v2.

galak commented on scripts/support/zephyr_flash_debug.py:
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
+
+
+class ZephyrBinaryRunner(abc.ABC):
+    '''Abstract superclass for binary runners (flashers, debuggers).'''
```
Can you add more docs about the theory/practice of how this class would be utilized.

mbolivar replied:
Sounds good.

mbolivar replied:
Sure; I've tried to improve this

galak reviewed (changes requested):
What about refactor this a little so that each of the runners are in their own files, and make it a bit more library so other script front ends might utilize the same bits.

I'm thinking about how we might write or modify sanitycheck to allow running tests built on a local board.

mbolivar commented:
> What about refactor this a little so that each of the runners are in their own files, and make it a bit more library so other script front ends might utilize the same bits.

Sure, that's fine. Can I do that on top of this series since it already starts out with them all in one file?

Commit:
scripts: zephyr_flash_debug: get bossac port from environment

Allow the environment variable BOSSAC_PORT to specify the device's
serial port, if present.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: zephyr_flash_debug: refactor into new "runner" package

Factor the classes which do the work into a new "runner" package. This
package has a core module where ZephyrBinaryRunner and common helpers
will live, and one file per subclass / runner front-end.

The top-level script, zephyr_flash_debug.py, still exists, but just
delegates its work to the core.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
build: remove flash/debug shell scripts

The functionality of the shell scripts for flashing and debugging has
now been replaced by zephyr_flash_debug.py. Remove the legacy scripts
as part of transitioning all of this to Python.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

lpereira commented on scripts/support/zephyr_flash_debug.py:
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
+
+    if system == 'Windows':
+        cflags |= subprocess.CREATE_NEW_PROCESS_GROUP
+    elif system in {'Linux', 'Darwin'}:
```
What's the value of `os.setsid` on Windows? Might be the case that, if it's available, use it.

lpereira commented on scripts/support/zephyr_flash_debug.py:
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
     if debug:
-        print(' '.join(cmd))
+        print(quote_sh_list(cmd))
     return subprocess.check_output(cmd)
```
Are we doing the shell escaping ourselves because of Windows? Does it have anything similar to `execv()` that can be used instead?

mbolivar replied:
This uses the standard Python module shlex behind the scenes. In addition to making this portable for Windows, this also means that the output can be copy and pasted and the shell tokenization wil remain what the command in list form intended.

mbolivar replied:
Looking at this again, I think you might be asking about the subprocess.check_output line, not quote_sh_list(). If that's right, I'm not sure what you mean by "doing the shell escaping ourselves". The calling conventions for subprocess routines recommend using a sequence for reasons explained here:

https://docs.python.org/3/library/subprocess.html#frequently-used-arguments

Given that cmd is a list, this looks very similar to execv(), so I think it's doing what you're looking for. Please LMK if not.

lpereira commented on scripts/support/zephyr_flash_debug.py:
```diff
@@ -38,105 +40,236 @@ def get_env_bool_or(env_var, default_value):
+        return self._parser_windows(cmd)
+
+    def _used_now_linux(self):
+        cmd = ['netstat', '-a', '-n', '-t']
```
`netstat` has been deprecated in Linux; my system at home doesn't even have it installed.  The `ss` utility from `iproute2` is the replacement. It might be the case both need to be tried.

mbolivar replied:
Hrm, OK. Use of netstat matches the existing shell script, which I'm trying to replicate fairly closely in this series (except of course where it needs to be changed for non-Linux targets).

Would you be OK with me fixing this to use ss in a follow-up? I can do it now if that's better, but it feels like the wrong series to do it in.

lpereira replied:
It can be done in a follow up patch, no problems.

mbolivar replied:
Thanks, I'll queue this up in the next series:

https://github.com/mbolivar/zephyr/commit/57e096f165d6c59affb42d719d69bb71fb51dc0c

andrewboie commented:
I concur with @galak I would like to see the arch or platform specific script logic separated out from the main script.

mbolivar commented:
v2 is posted now addressing comments from @galak. Changes from v1:

- **Delete all shell scripts** in a patch at the end
- Add a couple of debug= arguments to constructors where they were missing (not sure how this got past my flake8 test last time)
- Update the bossac runner to take a serial port from  the environment
- Add a patch which refactors the unified script into a shim on top of a library with individual files per runner
- Add more documentation and theory of ops to ZephyrBinaryRunner

mbolivar commented:
@andrewboie, done.

@lpereira, I have responded to your comments; please let me know if you'd like further changes along those lines.

tyrantlizard77 reviewed (approved):
approach taken here looks reasonable to me

mbolivar commented:
v3: rebased on top of https://github.com/zephyrproject-rtos/zephyr/pull/4502

galak reviewed (changes requested):
Can you move the refactor script/lib patch before adding the new classes.

Commit:
scripts: runner: add network port helper class

Some of the flashing scripts try to be clever about picking unused
ports. That's convenient for the user, so add a helper class to
runner.core to accomplish similar ends portably.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: runner: add debug support to nios.py

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: runner: add jlink.py, with debug and debugserver support

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: runner: add xtensa.py, with debug support

This is debug only, not debugserver.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

Commit:
scripts: runner: add qemu.py, a stand-in for QEMU

This, like the shell script, is just a placeholder.

Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>

mbolivar commented:
> Can you move the refactor script/lib patch before adding the new classes.

Done. Diff from the previous version is just deleting an extraneous import caught by flake8: https://hastebin.com/nehularafa.sql

mbolivar commented:
One more try. Botched the the refactor __init__.py by importing classes that didn't get added until later. Same end result, but intermediate commits are OK now.
